server/src/main/java/se/fortnox/reactivewizard/server/RxNettyServer.java

```diff
@@ -65,17 +75,50 @@ public void run() {
     }
 
     private void registerShutdownHook() {
-        Runtime.getRuntime().addShutdownHook(new Thread(() -> shutdownHook(config, server, connectionCounter)));
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> shutdownHook(config, server, eventLoopGroup, connectionCounter)));
+    }
+
+    public static void registerShutdownDependency(Runnable blockShutdownUntil) {
+        if (RxNettyServer.blockShutdownUntil != null && blockShutdownUntil != null) {
+            throw new IllegalStateException("Shutdown dependency is already registered");
+        }
+        RxNettyServer.blockShutdownUntil = blockShutdownUntil;
     }
 
-    static void shutdownHook(ServerConfig config, HttpServer server, ConnectionCounter connectionCounter) {
-        LOG.info("Shutdown requested. Will wait up to {} seconds...", config.getShutdownTimeoutSeconds());
-        server.shutdown();
+    static void shutdownHook(ServerConfig config, HttpServer server, EventLoopGroup eventLoopGroup, ConnectionCounter connectionCounter) {
+        LOG.info("Shutdown requested.");
+        awaitShutdownDependency();
+        LOG.info("Will wait up to {} seconds...", config.getShutdownTimeoutSeconds());
+        shutdownEventLoopGracefully(config, eventLoopGroup);
         if (!connectionCounter.awaitZero(config.getShutdownTimeoutSeconds(), TimeUnit.SECONDS)) {
             LOG.error("Shutdown proceeded while connection count was not zero: " + connectionCounter.getCount());
         }
         server.awaitShutdown();
         LOG.info("Shutdown complete");
     }
 
+    static void awaitShutdownDependency() {
+        if (blockShutdownUntil == null) {
+            return;
+        }
+
+        LOG.info("Wait for completion of shutdown dependency");
+        Thread thread = new Thread(blockShutdownUntil);
+        thread.start();
+        try {
+            thread.join();
```

Comment:
This join needs to have a timelimit that is the time left until config.getShutdownTimeoutSeconds() has been reached.